File: packages/sdk-go/agent.go
package whyops

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"sync"
)

type agentRegistry struct {
	mu             sync.Mutex
	cache          map[string]*AgentInfo
	apiKey         string
	proxyBaseURL   string
	analyseBaseURL string
	http           *httpClient
}

func newAgentRegistry(apiKey, proxyBaseURL, analyseBaseURL string) *agentRegistry {
	return &agentRegistry{
		cache:          make(map[string]*AgentInfo),
		apiKey:         apiKey,
		proxyBaseURL:   proxyBaseURL,
		analyseBaseURL: analyseBaseURL,
		http:           newHTTPClient(),
	}
}

func (r *agentRegistry) ensure(ctx context.Context, agentName string, metadata AgentMetadata) *AgentInfo {
	key := agentName + ":" + stableJSON(metadata)

	r.mu.Lock()
	if info, ok := r.cache[key]; ok {
		r.mu.Unlock()
		return info
	}
	r.mu.Unlock()

	info := r.init(ctx, agentName, metadata)
	if info != nil {
		r.mu.Lock()
		r.cache[key] = info
		r.mu.Unlock()
	}
	return info
}

func (r *agentRegistry) init(ctx context.Context, agentName string, metadata AgentMetadata) *AgentInfo {
	// Ensure tools is always an array (backend requires it)
	tools := metadata.Tools
	if tools == nil {
		tools = []AgentTool{}
	}
	body := map[string]any{
		"agentName": agentName,
		"metadata": map[string]any{
			"systemPrompt": metadata.SystemPrompt,
			"description":  metadata.Description,
			"tools":        tools,
		},
	}
	headers := map[string]string{"Authorization": "Bearer " + r.apiKey}

	urls := []string{
		r.analyseBaseURL + "/entities/init",
		r.proxyBaseURL + "/v1/agents/init",
	}

	for _, url := range urls {
		status, data, err := r.http.post(ctx, url, body, headers)
		if err != nil || status < 200 || status >= 300 {
			continue
		}

		var resp struct {
			AgentID        string `json:"agentId"`
			AgentVersionID string `json:"agentVersionId"`
			Status         string `json:"status"`
			VersionHash    string `json:"versionHash"`
		}
		if err := json.Unmarshal(data, &resp); err != nil || resp.AgentID == "" {
			continue
		}

		return &AgentInfo{
			AgentID:        resp.AgentID,
			AgentVersionID: resp.AgentVersionID,
			Status:         resp.Status,
			VersionHash:    resp.VersionHash,
		}
	}

	log.Println("[whyops] agent init failed — continuing without registration")
	return nil
}

func stableJSON(v any) string {
	b, err := json.Marshal(v)
	if err != nil {
		return fmt.Sprintf("%v", v)
	}
	return string(b)
}

File: packages/sdk-python/whyops/agent.py
from __future__ import annotations

import json
import sys
from typing import Optional

from ._config import (
    ENDPOINT_AGENT_INIT_FALLBACK,
    ENDPOINT_AGENT_INIT_PRIMARY,
    LOG_PREFIX,
)
from .http import post_async, post_sync
from .types import AgentInfo, AgentMetadata


def _stable_key(agent_name: str, metadata: AgentMetadata) -> str:
    return f"{agent_name}:{json.dumps(metadata, sort_keys=True)}"


class AgentRegistry:
    def __init__(self, api_key: str, proxy_base_url: str, analyse_base_url: str) -> None:
        self._api_key = api_key
        self._proxy_base_url = proxy_base_url.rstrip("/")
        self._analyse_base_url = analyse_base_url.rstrip("/")
        self._cache: dict[str, AgentInfo] = {}

    def _urls(self) -> list[str]:
        return [
            f"{self._analyse_base_url}{ENDPOINT_AGENT_INIT_PRIMARY}",
            f"{self._proxy_base_url}{ENDPOINT_AGENT_INIT_FALLBACK}",
        ]

    def _headers(self) -> dict[str, str]:
        return {"Authorization": f"Bearer {self._api_key}"}

    def _body(self, agent_name: str, metadata: AgentMetadata) -> dict:
        m = {"tools": [], **dict(metadata)}
        return {"agentName": agent_name, "metadata": m}

    # ── Sync ─────────────────────────────────────────────────────────────────

    def ensure_sync(self, agent_name: str, metadata: AgentMetadata) -> Optional[AgentInfo]:
        key = _stable_key(agent_name, metadata)
        if key in self._cache:
            return self._cache[key]
        info = self._init_sync(agent_name, metadata)
        if info:
            self._cache[key] = info
        return info

    def _init_sync(self, agent_name: str, metadata: AgentMetadata) -> Optional[AgentInfo]:
        body = self._body(agent_name, metadata)
        headers = self._headers()
        for url in self._urls():
            try:
                r = post_sync(url, body, headers)
                if r.is_success:
                    data = r.json()
                    if data.get("agentId"):
                        return AgentInfo(
                            agentId=data["agentId"],
                            agentVersionId=data["agentVersionId"],
                            status=data["status"],
                            versionHash=data["versionHash"],
                        )
            except Exception:
                pass
        print(f"{LOG_PREFIX} agent init failed — continuing without registration", file=sys.stderr)
        return None

    # ── Async ────────────────────────────────────────────────────────────────

    async def ensure_async(self, agent_name: str, metadata: AgentMetadata) -> Optional[AgentInfo]:
        key = _stable_key(agent_name, metadata)
        if key in self._cache:
            return self._cache[key]
        info = await self._init_async(agent_name, metadata)
        if info:
            self._cache[key] = info
        return info

    async def _init_async(self, agent_name: str, metadata: AgentMetadata) -> Optional[AgentInfo]:
        body = self._body(agent_name, metadata)
        headers = self._headers()
        for url in self._urls():
            try:
                r = await post_async(url, body, headers)
                if r.is_success:
                    data = r.json()
                    if data.get("agentId"):
                        return AgentInfo(
                            agentId=data["agentId"],
                            agentVersionId=data["agentVersionId"],
                            status=data["status"],
                            versionHash=data["versionHash"],
                        )
            except Exception:
                pass
        print(f"{LOG_PREFIX} agent init failed — continuing without registration", file=sys.stderr)
        return None
